### ISSUE TITLE
Question: Is the package able to deal with non-constant coefficients

### ISSUE THREAD

**Issue by ItayXD** [zwicker-group/py-pde#22]
Hi,
I didn't see it mentioned anywhere and it doesn't appear in any of the examples.
Can I define a PDE of the form:
<img src="https://render.githubusercontent.com/render/math?math=\frac{\partial P}{\partial t} = D(x) \cdot \frac{\partial^2 P}{{\partial x}^2}">
where D(x) could be a polynomial of x for example

Thanks
Itay

**Comment by david-zwicker** [comment]
This is certainly possible. The simplest way would be to use the [`PDE` class](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.html#pde.pdes.pde.PDE),

```python
from pde import PDE
eq = PDE({'P': '(P - P**2) * laplace(P)'})
```
where I used `D(x) = x - x**2` as a concrete example. Alternatively, you may [implement your own pde class](https://py-pde.readthedocs.io/en/latest/manual/advanced_usage.html#custom-pde-classes).

See [`pde_custom_expression.py`](https://github.com/zwicker-group/py-pde/blob/master/examples/pde_custom_expression.py) and [`pde_custom_class.py`](https://github.com/zwicker-group/py-pde/blob/master/examples/pde_custom_class.py) in the examples directory for respective implementations.

**Comment by ItayXD** [comment]
I may be missing something, but it seems that you actually used `D(P)=P - P**2` rather than `D(x) = x - x**2`, meaning the coefficient depends on the variable being solved for and not x (which P is being differentiated with respect to)

**Comment by david-zwicker** [comment]
Ohh, I wasn't reading your question carefully enough – sorry about this. The current implementation of the `PDE` class does not allow spatially varying coefficients. However, it is not too difficult to achieve this with a custom class:

```python
import pde


class HeterogeneousDiffusionPDE(pde.PDEBase):

    def evolution_rate(self, state, t=0):
        """ implement the python version of the evolution equation """
        assert state.grid.dim == 1  # implementation only works for 1d problems
        cell_coords = state.grid.cell_coords[:, 0]  # this extracts the x-coordinates of all cells
        coeff = (cell_coords - cell_coords**2)  # this calculates the pre-factor
        return state.laplace('natural') * coeff


grid = pde.CartesianGrid([[0, 1]], 32)  # generate grid
state = pde.ScalarField.random_normal(grid, 0.1, 0.9)  # generate initial condition

eq = HeterogeneousDiffusionPDE()  # define the pde
result = eq.solve(state, t_range=1, dt=1e-4)
result.plot()
```

A numba-compiled implementation to accelerate the computation can be created similarly. I'll think about extending the `PDE` class, but it might be a bit challenging and thus not happen soon. However, pull requests in this direction are welcome!

**Comment by david-zwicker** [comment]
I took you suggestion as a challenge and implemented an extended version of the `PDE` class that can deal with spatially dependent coefficients. The implementation is not very tested, yet, but it seems to work. I added [an example that displays how it works](https://github.com/zwicker-group/py-pde/blob/master/examples/pde_heterogeneous_diffusion.py), but it really comes down to just using the spatial coordinates as variables in the expressions.
